### DISCUSSION TITLE
Add multiple things (add another)

### DISCUSSION THREAD

**Discussion by isobelseacombe** [UKHomeOffice/design-system#353] [General]
### What exists already?
Home Office Design system guidance: [https://design.homeoffice.gov.uk/patterns/add-multiple-things](url)
MoJ pattern: [https://design-patterns.service.justice.gov.uk/patterns/add-to-a-list/](url)
MoJ component: [https://design-patterns.service.justice.gov.uk/components/add-another/](url)
GOV.UK backlog discussion: [https://github.com/alphagov/govuk-design-system-backlog/issues/21](url)

The existing pattern on the Home Office design system works well for public facing forms based services. When working on internal products where our users are adding lots of information, we want to make the user experience as efficient as possible so that they can get back to their other operational duties.

### We're looking for examples of designs that address:

- Adding multiple small pieces of information (short task flow)
- Adding detailed information about multiple things (long task flow)
- Adding multiple things as a component part of a longer form (part of a much longer task flow)

### Challenges
- If a user is taken through loops to add multiple items to a form, then do you wait until the end to show them what information they have entered?
- If a user is allowed to add multiple things on one page then how do we handle edits and going back and forth to make any changes that are needed?

**Comment by isobelseacombe** [comment]
https://design-system.dwp.gov.uk/patterns/add-another-thing

**Reply by guermanc** [reply]
This has some well-presented information in regards to content and accessibility that we may want to include.

**Comment by guermanc** [comment]
With so many patterns for adding multiple things already in existence, what would we be hoping to achieve by creating our own implementation?

**Comment by emilyvp** [comment]
### 'Add multiple things' research

**Overview** 

In our service we have implemented the ‘Add multiple things’ pattern and adapted it slightly to allow users to upload multiple documents. This is a public service that sits on a Home Office domain. 

When adding documents, we ask users to assign the document category to the document they are uploading from the ‘Select’ document component. 

In a previous iteration of the design, users had the ability to add multiple documents, with the added documents appearing in a table below. From user research, we identified the following issues: 

- it was unclear to users that multiple documents could be added. It wasn’t clear that the ‘Add this document’ button created an entry in the table below 
- the document table was below the line not immediately visible  
- users ended up in a confused flow/ dead end where they would expect to see the added document on the subsequent screen and would attempt to re-add the same document – being confused as to how they could add multiple documents 
- users missed the document type selection step – this step sat below the file upload step 

_Previous screen iteration_
![1  Upload screen](https://user-images.githubusercontent.com/55453519/219074835-04eaf4ed-dc2a-450b-928e-ec00184240f6.png)

Following these insights the following design changes were made: 

- the user can upload multiple documents one at a time, checking the documents on a subsequent screen and can add more if required, in a add multiple documents ‘loop’
- the user selects the document type before uploading a file 

_Revised design with add multiple documents loop_
![2  Upload loop](https://user-images.githubusercontent.com/55453519/219120380-6f625f79-4e7b-454e-a348-74810b938b90.png)


**Findings from user research:** 

- the improved design enabled users to upload multiple documents without encountering issues 
- displaying the uploaded documents on a subsequent screen reduces the cognitive load for the user 
- users became familiar with the 'loop' of categorising and adding more documents and were reassured when they could see them being added to the list each time. 
- asking the user to input additional information relating to the document they are uploading prior to the upload document step worked better 


**Additional findings regarding length of task flows and the ‘Check your answers’ pattern** 

The upload documents task previously sat in a long task flow where users had to complete 2 additional tasks, with a final ‘Check your answers’ screen at the end of the 3 steps.  

We found that splitting the tasks up into short task flows, with corresponding ‘Check’ screens for each task, tested much better with users - avoiding issues and confusion, allowing the user to have more flexibility for each step/ task. 

We updated the ‘Task list’ screen to include 3 tasks as opposed to 1 long task. 

**Proposal** 

Ideas for updates to the pattern based on the research: 

- specific guidance for uploading multiple documents  
- recommendation for splitting tasks up where possible, using a ‘Check screen’ for short tasks, rather than one long task with multiple steps and a final ‘Check screen’ 
- updating the list of services where this pattern is used
